lib/src/rpc.ml

```diff
@@ -24,3 +29,27 @@ let index_put =
   let bin_query = [%bin_type_class: string * string * string list]
   and bin_response = [%bin_type_class: unit] in
   Async.Rpc.Rpc.create ~name:"put_index" ~version:0 ~bin_query ~bin_response
+
+let metadata_put =
+  let bin_query = [%bin_type_class: string * string]
+  and bin_response = [%bin_type_class: unit] in
+  Async.Rpc.Rpc.create ~name:"put_metadata" ~version:0 ~bin_query ~bin_response
+
+let decode_block_get = function
```

Comment:
This seems to be a simple state machine that accepts the executable bit and file contents in sequence, but the code is still non-trivial, so it's worth writing down what it does and why.

Reply:
By the way, since it's pretty messy from the point of view of types (they don't really help us much here), why not just use a stream of strings and have a convention that the first string represents the boolean that can be parsed with something basic like `Bool.of_string_exn`? I bet the resulting code would be much nicer, safer and also faster.